test/hotspot/jtreg/compiler/locks/TestNestedSynchronize.java

```diff
@@ -23,216 +23,96 @@
- * @run main/othervm -XX:CompileCommand=compileonly,compiler.locks.TestNestedSynchronize::test
+ * @bug 8322996 8324839 8325467
+ * @summary Ensure C2 can compile deeply nested synchronize statements.
+ *          Exercises C2 register masks, in particular. We incrementally
+ *          increase the level of nesting (up to 100) to trigger potential edge
+ *          cases.
+ * @modules java.base/jdk.internal.misc
+ * @library /test/lib /
+ * @run main/othervm -XX:CompileCommand=compileonly,Test::test*
  *                   -Xcomp
+ *                   -XX:+UnlockDiagnosticVMOptions
+ *                   -XX:+AbortVMOnCompilationFailure
  *                   compiler.locks.TestNestedSynchronize
  */
 
 package compiler.locks;
 
+import compiler.lib.compile_framework.*;
+import java.util.LinkedList;
+
 public class TestNestedSynchronize {
+    static int min = 1;
+    static int max = 100;
+    static String test_class_name = "Test";
+    static String test_method_name = "test";
 
-    public static void main(String[] args) {
-        test();
+    // The below method generates a program of the form:
+    //
+    // public class Test {
+    //     public static void test1() {
+    //         synchronized (Test.class) {
+    //         }
+    //     }
+    //
+    //     public static void test2() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         }
+    //         }
+    //     }
+    //
+    //     public static void test3() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         }
+    //         }
+    //         }
+    //     }
+    //
+    //     ...
+    //
+    //     public static void test100() {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         synchronized (Test.class) {
+    //         ...
+    //         }
+    //         }
+    //         }
+    //     }
+    // }
+    //
+    // The above is a massive program. Therefore, we do not directly inline the
+    // program in TestNestedSynchronize and instead compile and run it via the
+    // CompileFramework.
+    public static String generate_test() {
+        LinkedList<String> acc = new LinkedList<String>();
+        for (int i = min; i <= max; i++) {
+            LinkedList<String> method = new LinkedList<String>();
+            for (int j = 0; j < i; j++) {
+                method.addFirst(String.format(
+                    "        synchronized (%s.class) {", test_class_name));
+                method.addLast("        }");
+            }
+            method.addFirst(String.format(
+                "    public static void %s%d() {", test_method_name, i));
+            method.addLast("    }");
+            acc.addAll(method);
+        }
+        acc.addFirst(String.format("public class %s {", test_class_name));
+        acc.addLast("}");
+        return String.join("\n", acc);
```

Comment:
Nit: What prevented you from generating it with a `ArrayList`, and just only append at the end?

In fact, that would allow you to use a StringBuilder directly.

And you could format the `synchronized` string only once. That might be even more efficient.

You can also leave it, this is really a nit :)

Reply:
Nothing prevented it, it just felt more natural doing it in this way. I guess a habit from doing a lot of functional programming :slightly_smiling_face:  It should have the same time complexity now as using a `StringBuilder` (linear), so I'd prefer leaving it as is unless you strongly object!